New dashboard/folder url structure

marefr opened the pull request:
Fixes #7883, #10419, #10579 and #10408

marefr committed:
db: add new column uid to the dashboard table. #7883

marefr committed:
db: add migrations for generating uid for existing dashboards. #7883

marefr committed:
db: add migrations for creating a unique index for uid. #7883

marefr committed:
dashboard: generate and include uid in dashboard model. #7883

marefr committed:
dashboard: fix failing test. #7883

marefr committed:
dashboards: extract short uid generator to util package. #7883

marefr committed:
dashboards: return uid in response to creating/updating a dashboard. #7883

marefr committed:
dashboards: add support for retrieving a dashboard by uid

Introduces new url in api /dashboards/<uid> for fetching dashboard by unique id
Leave the old dashboard by slug url /dashboards/db/<slug> for backward
compatibility and for supporting fallback
WIP for #7883

marefr committed:
dashboard: change unique index for uid to include org_id

Make the max length of uid longer in case we want to change it later
#7883

marefr committed:
dashboards: api for retrieving uid by slug. #7883

marefr committed:
dashboards: add url property to dashboard meta and search api responses

#7883

bergquist committed:
Drops unique index orgid_slug from dashboards.

bergquist committed:
removes uniqnes check on slug when saving dashboards

DanCech commented on pkg/util/shortid_generator.go:
```diff
@@ -0,0 +1,24 @@
+			}
+		}
+	}
+
```
Why is this necessary? I don't see anything in `shortid.Generate` that is expected to fail, and the library itself provides `shortid.MustGenerate` which panics on failure, so I don't think the author expects it to fail either

marefr replied:
That's a really good question :) Late coding and not thinking about what possible errors the lib could generate. I'll fix. Thanks for the extra eye :+1:

marefr committed:
util: remove retry logic in shortid_generator

Use shortid.MustGenerate() instead of shortid.Generate(). Instead
of returning errors it will panic.

marefr committed:
dashboards: revert adding api for retrieving uid by slug

Since we're already have possibility to get a dashboard by slug
it makes little sense to have a separate endpoint in api for
retrieving uid by slug.
#7883

marefr committed:
dashboards: return url in response to save dashboard. #7883

marefr committed:
dashboards: ensure that uid is returned from getSaveModelClone

Without this the uid will not be sent to the backend when saving a
dashboard.
#7883

marefr committed:
dashboards: add new default frontend route for loading a dashboard

New default route /d/<uid>/<slug of dashboard title> where dashboard
are loaded by unique identifier.
If old route /dashboard/db/<slug of dashboard tile> are used,
try to lookup dashboard by slug and redirect to new default route.
 #7883

marefr committed:
dashboards: when saving dashboard redirect if url changes

Redirect if new dashboard created or existing url changes.
#7883

marefr committed:
dashboards: use new *url* prop from dashboard search for linking to dashboards

#7883

marefr committed:
dashboards: fix links to recently viewed and starred dashboards

Use new property url from dashboard search for linking to dashboards
#7883

marefr committed:
dashboards: add new default frontend route for rendering a dashboard panel

New default route /d-solo/<uid>/<slug of dashboard title> where dashboard
panel are rendered by unique identifier and panel identifier.
If old route /dashboard-solo/db/<slug of dashboard tile> are used,
try to lookup dashboard by slug and redirect to new default route.
Change url logic for sharing a panel so that the new default route
for rendering a dashboard panel are used.
 #7883

xlson commented on pkg/api/dashboard.go:
```diff
@@ -44,7 +44,7 @@ func dashboardGuardianResponse(err error) Response {
 
 func GetDashboard(c *middleware.Context) Response {
-	dash, rsp := getDashboardHelper(c.OrgId, c.Params(":slug"), 0)
+	dash, rsp := getDashboardHelper(c.OrgId, c.Params(":slug"), 0, c.Params(":uid"))
```
May want to create two different GetDashboard functions depending on if we want to get by slug or by uid so we dont have to handle both cases in the same func.

xlson commented on pkg/api/dashboard.go:
```diff
@@ -101,6 +101,12 @@ func GetDashboard(c *middleware.Context) Response {
 	}
 
+	if dash.IsFolder {
+		meta.Url = m.GetFolderUrl(dash.Uid, dash.Slug)
```
Should do this in 5.1 and keep folder urls as is.

marefr replied:
Fixed new folder url in frontend_step based on uid, i.e. `/dashboards/f/<uid>/<slug>`.

DanCech replied:
I see this repeated here and in PostDashboard, should it be a function on dashboard instead?

marefr replied:
Yep, you're right. We should fix this.

xlson commented on pkg/api/dashboard.go:
```diff
@@ -124,16 +130,23 @@ func getUserLogin(userId int64) string {
 
 func DeleteDashboard(c *middleware.Context) Response {
-	dash, rsp := getDashboardHelper(c.OrgId, c.Params(":slug"), 0)
+	dash, rsp := getDashboardHelper(c.OrgId, c.Params(":slug"), 0, "")
```
This wont work when we remove slug uniqueness.

marefr replied:
Fixed in frontend_step2 branch so will be fixed when merging in that

DanCech replied:
Just a comment, having multiple branches going at the same time seems to be causing a lot of confusion. Is there a reason we're not just doing the dev in a single branch?

marefr replied:
At first we were unsure if we was going to be able to include all the needed changes in beta 1. So therefore we decided to split up the initial work to have the possibility to merge the initial work to beta 1. 

Second, after the initial work we had different tasks that could be divided between several people with possibility to work in parallel on multiple branches. Having multiple branches we could review them separately and decide to merge some of the work and include it in beta 1. 

One alternative could have been to use one branch and cherry pick commits to beta 1 branch, but then there's always a risk that you cannot cherry pick one commit because it's dependent on another that we may not want to hit beta 1.

The idea is still to review the branches using separate PR's and then merge them to this branch one at a time. Hope this make things more clearer,

Right now it looks promising for getting almost everything included in the beta 1.

xlson commented on pkg/api/dashboard.go:
```diff
@@ -393,7 +406,7 @@ func CalculateDashboardDiff(c *middleware.Context, apiOptions dtos.CalculateDiff
 // RestoreDashboardVersion restores a dashboard to the given version.
 func RestoreDashboardVersion(c *middleware.Context, apiCmd dtos.RestoreDashboardVersionCommand) Response {
-	dash, rsp := getDashboardHelper(c.OrgId, "", c.ParamsInt64(":dashboardId"))
+	dash, rsp := getDashboardHelper(c.OrgId, "", c.ParamsInt64(":dashboardId"), "")
```
This wont work when we remove slug uniqueness.

marefr replied:
That should still work since we're using dashboard id for retrieving the dashboard.

xlson reviewed (changes requested):
Looked over the changes together with Carl and found some things we want to change.

bergquist committed:
ensure dashboard title is unique in folder

marefr committed:
dashboards: redirect from old url used to load dashboard to new url

If legacy backend routes (/dashboard/db/<slug> and /dashboard-solo/db/<slug>)
are requested we try to redirect to new routes with a 301 Moved Permanently
 #7883

marefr committed:
dashboards: make scripted dashboards work using the old legacy urls

Scripted dashboards are still requested from /dashboard/script/scripted.js
#7883

bergquist committed:
Verifies requirement of id in dashboards.

bergquist committed:
gofmt...

bergquist committed:
Merge remote-tracking branch 'origin/7883_frontend_step' into 7883_backend

* origin/7883_frontend_step:
  dashboards: make scripted dashboards work using the old legacy urls
  dashboards: redirect from old url used to load dashboard to new url
  dashboards: add new default frontend route for rendering a dashboard panel
  dashboards: fix links to recently viewed and starred dashboards
  dashboards: use new *url* prop from dashboard search for linking to dashboards
  dashboards: when saving dashboard redirect if url changes
  dashboards: add new default frontend route for loading a dashboard
  dashboards: return url in response to save dashboard. #7883

marefr committed:
dashboards: new route for deleting dashboards by uid

marefr committed:
dashboards: add validation to delete dashboard by slug

Validates that there are only one folder/dashboard having that slug,
otherwise returns 412 Precondition Failed

marefr committed:
folders: change the front end route for browsing folders

Change the front end route for folders to /dashboards/f/<uid>/<slug of folder title>.
Use new route for deleting dashboard/folder by uid.
Retrieve dashboard/folder by uid when moving or deleting dashboards/folders.

marefr committed:
dashboards: remove slug property in dashboard search responses

Removes slug property in dashboard search responses since this property isn't needed
anymore and it haven't been released to any stable release.

torkelo committed:
fix: use replace when redirecting to new url

bergquist committed:
retry uid generation

marefr committed:
dashboards: fix updating folder so that correct url is returned

marefr committed:
dashboards: when restoring a dashboard to an older version, set current uid

marefr committed:
dashboards: update dashboard/folder url if browser url is not the same as from backend

marefr committed:
Merge pull request #10695 from grafana/7883_frontend_step

WIP: Dashboard & Persistent urls - Frontend Step

marefr committed:
dashboard: refactor logic for retrieving url for folder/dashboard

bergquist committed:
Merge pull request #10694 from grafana/7883_backend

backend for new url structure.

marefr committed:
Merge remote-tracking branch 'upstream/master' into 7883_new_url_structure

marefr committed:
Merge branch '7883_new_url_structure' into 7883_frontend_step2

bergquist committed:
changes dashboard url in alertlist

marefr committed:
folders: changes needed due to merge

marefr committed:
route params from angular to view store should be updated on routeChangeSuccess

bergquist committed:
alert: use new url format

bergquist committed:
make it easier for dashboards to generate ur;

marefr committed:
Merge pull request #10706 from grafana/7883_frontend_step2

WIP: Dashboard & Persistent urls - Frontend Step 2

bergquist committed:
register handler for get dashboards by slug

bergquist committed:
Merge pull request #10715 from grafana/7883_alerting

Implement new url format in alerting

marefr committed:
fix for dashboard/folder url's when having a sub path in root_url config

marefr committed:
Merge branch '7883_new_url_structure' of github.com:grafana/grafana into 7883_new_url_structure

marefr committed:
Merge branch 'master' into 7883_new_url_structure

DanCech commented:
Nice work @marefr 👍

marefr commented:
There's an issue with validation of titles for folder/dashboard in the frontend. We have some validation logic in validation_srv.ts that needs to be removed or updated.
